app/javascript/crayons/MultiSelectAutocomplete/MultiSelectAutocomplete.jsx

```diff
@@ -0,0 +1,316 @@
+        }
+        break;
+      default:
+        if (!ALLOWED_CHARS_REGEX.test(e.key)) {
+          e.preventDefault();
+        }
+    }
+  };
+
+  // If there is a previous selection, then pop it into edit mode
+  const editPreviousSelectionIfExists = () => {
+    if (selectedItems.length > 0 && inputPosition !== 0) {
+      const nextEditIndex =
+        inputPosition !== null ? inputPosition - 1 : selectedItems.length - 1;
+
+      const item = selectedItems[nextEditIndex];
+      deselectItem(item);
+      enterEditState(item, nextEditIndex);
+    }
+  };
+
+  const selectItem = ({
+    selectedItem,
+    nextInputValue = '',
+    focusInput = true,
+  }) => {
+    // If an item was edited, we want to keep it in the same position in the list
+    const insertIndex = inputPosition ? inputPosition : selectedItems.length;
+    const newSelections = [
+      ...selectedItems.slice(0, insertIndex),
+      selectedItem,
+      ...selectedItems.slice(insertIndex),
+    ];
+
+    // We update the hidden selected items list, so additions are announced to screen reader users
+    const listItem = document.createElement('li');
+    listItem.innerText = selectedItem;
+    selectedItemsRef.current.appendChild(listItem);
+
+    exitEditState(nextInputValue);
+    setSelectedItems(newSelections);
+    // Clear the currently displayed suggestions & active index
+    setSuggestions([]);
+    setActiveDescendentIndex(null);
+    // Clear the text input
+    const { current: input } = inputRef;
+    input.value = nextInputValue;
+    focusInput && input.focus();
+  };
+
+  const deselectItem = (deselectedItem) => {
+    setSelectedItems(selectedItems.filter((item) => item !== deselectedItem));
+
+    // We also update the hidden selected items list, so removals are announced to screen reader users
+    selectedItemsRef.current.querySelectorAll('li').forEach((selectionNode) => {
+      if (selectionNode.innerText === deselectedItem) {
+        selectionNode.remove();
+      }
+    });
+  };
+
+  const allSelectedItemElements = selectedItems.map((item, index) => (
+    <li key={item} className="w-max">
+      <div role="group" aria-label={item} className="flex mr-1 mb-1 w-max">
+        <Button
+          variant="secondary"
+          className="c-autocomplete--multi__selected p-1 cursor-text"
+          aria-label={`Edit ${item}`}
+          onClick={() => enterEditState(item, index)}
+        >
+          {item}
+        </Button>
+        <Button
+          variant="secondary"
+          className="c-autocomplete--multi__selected p-1"
+          aria-label={`Remove ${item}`}
+          onClick={() => deselectItem(item)}
+        >
+          <Icon src={Close} />
+        </Button>
+      </div>
+    </li>
+  ));
+
+  // When a user edits a tag, we need to move the input inside the selected items
+  const splitSelectionsAt =
+    inputPosition !== null ? inputPosition : selectedItems.length;
+
+  const input = (
+    <li className="self-center">
```

Comment:
It is weird to have the input in a list item, inside the selected items list, but this makes it possible for us to wrap selections to a new line easily etc.